### PR TITLE
Add a `_freeze` / `_unfreeze` API

### PR DESCRIPTION
**Pull request by s1monw**
This commit adds a rest endpoint for freezing and unfreezing an index.
Among other cleanups mainly fixing an issue accessing package private APIs
from a plugin that got caught by integration tests this change also adds
documentation for frozen indices.
Note: frozen indices are marked as `beta` and available as a basic feature.

Relates to #34352

### CODE REVIEW

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
to be writable at the same time and have different access patterns over time. For example indices in the time-series or logging use-case
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
In order to keep indices available and queryable for a longer period but at the same time reduce their hardware requirements they can be transitioned
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
into a frozen state. Once an index is frozen, all of it's transient shard memory (aside from mappings and analyzers)
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
is moved to persistent storage. This allows for a much higher disk to heap storage ratio on individual nodes. Once an index is
```

File: docs/reference/modules/threadpool.asciidoc
Comment (clintongormley) [inline docs/reference/modules/threadpool.asciidoc]:
```suggestion
[[search-throttled]]`search_throttled`::
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
frozen, it is made read-only and drops its transient data structures from memory.  These will need to be reloaded on demand (and subsequently dropped) for each search request that targets the frozen index.  A search request that hits
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
one or more frozen shards will be executed on a throttled threadpool that ensures that we never search more than
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
`N` (`1` by default) searches concurrently (see <<search-throttled>>). This protects nodes from exceeding the available memory due to incoming search requests.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
In contrast to ordinary open indices, frozen indices are expected to execute slowly and are not designed for high query load. Parallelism is
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
execution on a per shard level. Depending on the data in an index, a frozen index may execute searches in the seconds to minutes range, when the same index in an unfrozen state may execute the same search request in milliseconds.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
Since frozen indices provide a much higher disk to heap ratio at the expense of search latency,  it is advisable to allocate frozen indices to
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
dedicated nodes to prevent searches on frozen indices influencing traffic on low latency nodes. There is significant overhead in loading
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
data-structures on demand which can cause page-faults and garbage collections, which further slow down query execution.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
Since indices that are eligible for freezing are unlikely to change in the future, disk space can be optimized as described in <<tune-for-disk-usage>>.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
The freeze and unfreeze index APIs are used to freeze and unfreeze an index.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
A frozen index has almost no overhead on the cluster (except
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
When a frozen index is unfrozen, then index will go through the normal recovery process and become writeable again.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
 amount of time and causes the cluster to go red until the primaries are allocated again. This limitation might be removed in the future
```
```suggestion
 amount of time and causes the cluster to go red until the primaries are allocated again. This limitation might be removed in the future.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
limited by the number of threads in the <<search-throttled>> threadpool,  which is `1` by default. 
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
Search requests will not be executed against frozen indices by default, even if a frozen index is named explicitly. Th is is ```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
to prevent accidental slowdowns by targeting a frozen index by mistake.. To include frozen indices a search request must be executed with
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
the query parameter `ignore_throttled=false`.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
While frozen indices are slow to search, they can be pre-filtered efficiently. The request parameter `pre_filter_shard_size` specifies
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
a threshold that, when exceeded, will enforce a round-trip to pre-filter search shards that cannot possibly match.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
This filter phase can limit the number of shards significantly. For instance, if a date range filter is applied, then all indices (frozen or unfrozen) that do not contain documents within the date range can be skipped efficiently.
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
```

File: docs/reference/frozen-indices.asciidoc
Comment (clintongormley) [inline docs/reference/frozen-indices.asciidoc]:
```suggestion
The default value for `pre_filter_shard_size` is `128` but it's recommended to set it to `1` when searching frozen indices. There is no
```

**Review by clintongormley — APPROVED** [review]
LGTM once changes are merged

File: docs/reference/frozen-indices.asciidoc
Comment (s1monw) [inline docs/reference/frozen-indices.asciidoc]: @lcawl any pointers how and where we wire this in are appreciated!

Comment (lcawl) [reply docs/reference/frozen-indices.asciidoc]: In my opinion, this content can all stay in the Elasticsearch Reference.  To that end, I submitted a couple of commits to make it work.  You can preview it here: https://doc-review-63bf6.firebaseapp.com/es2/frozen-indices.html

File: docs/reference/frozen-indices.asciidoc
Comment (lcawl) [inline docs/reference/frozen-indices.asciidoc]: In my opinion, this page "Freezing and unfreezing..." page is no longer needed after we move this content to the API reference pages (e.g. https://doc-review-63bf6.firebaseapp.com/es2/freeze-index-api.html and https://doc-review-63bf6.firebaseapp.com/es2/unfreeze-index-api.html)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java
Comment (tlrx) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java]: Nit: Index.EMPTY_ARRAY?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java
Comment (tlrx) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java]: Nit: extra line

File: x-pack/plugin/src/test/resources/rest-api-spec/test/indices.freeze/10_basic.yml
Comment (tlrx) [inline x-pack/plugin/src/test/resources/rest-api-spec/test/indices.freeze/10_basic.yml]: Can we change the reason here to mention Freeze/Unfreeze API?

File: x-pack/plugin/src/test/resources/rest-api-spec/test/indices.freeze/10_basic.yml
Comment (tlrx) [inline x-pack/plugin/src/test/resources/rest-api-spec/test/indices.freeze/10_basic.yml]: Same here

**Review by tlrx — APPROVED** [review]
LGTM, I left only minor comments.

File: x-pack/plugin/src/test/resources/rest-api-spec/test/indices.freeze/10_basic.yml
Comment (s1monw) [reply x-pack/plugin/src/test/resources/rest-api-spec/test/indices.freeze/10_basic.yml]: well that is the reason. We will have the same test in 6.x but with types and a different reason?

Comment (tlrx) [reply x-pack/plugin/src/test/resources/rest-api-spec/test/indices.freeze/10_basic.yml]: Ok

**Review by bleskes — COMMENTED** [review]
Looks great. I left some comments and questions.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java
Comment (s1monw) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java]: not by default. we have a test for this. if you set the indices option to include closed ones it does.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java
Comment (s1monw) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java]: it has been done twice. this one is enough https://github.com/elastic/elasticsearch/pull/35592/files#diff-be7d916f446b8daf34f954a4ad1e8271L101

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java
Comment (bleskes) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java]: I missed the fact we don't resolve closed indices by default. Fair enough. Sorry for the noise.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java
Comment (s1monw) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java]: fixed it here https://github.com/elastic/elasticsearch/pull/35592/commits/9648d7172e8a4af2abd95334ddeffafd4a6d4b5c

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java
Comment (s1monw) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java]: no worries!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java
Comment (s1monw) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java]: I pushed changes

**Pull request merged**
